### ISSUE TITLE
Unable to Set External Store in client-go

### ISSUE THREAD

**Issue by jeevanragula** [kubernetes/kubernetes#114659]
### What would you like to be added?

In shared_informer.go, when the resources are fetched from API Server, they are always cached in Memory and there is no mechanism to control that store. 

```
func NewSharedIndexInformer(lw ListerWatcher, exampleObject runtime.Object, defaultEventHandlerResyncPeriod time.Duration, indexers Indexers) SharedIndexInformer {
	realClock := &clock.RealClock{}
	sharedIndexInformer := &sharedIndexInformer{
		processor:                       &sharedProcessor{clock: realClock},
		indexer:                         NewIndexer(DeletionHandlingMetaNamespaceKeyFunc, indexers),
		listerWatcher:                   lw,
		objectType:                      exampleObject,
		resyncCheckPeriod:               defaultEventHandlerResyncPeriod,
		defaultEventHandlerResyncPeriod: defaultEventHandlerResyncPeriod,
		cacheMutationDetector:           NewCacheMutationDetector(fmt.Sprintf("%T", exampleObject)),
		clock:                           realClock,
	}
	return sharedIndexInformer
}
```
It is better to have a provision to supply the store as well something like below.
I would like to write the resource information to disk instead of memory to reduce the memory usage.

```
func (s *sharedIndexInformer) SetStore(store ThreadSafeStore) error {
	s.startedLock.Lock()
	defer s.startedLock.Unlock()

	if s.started {
		return fmt.Errorf("informer has already started")
	}
	s.indexer = NewIndexerWithStore(DeletionHandlingMetaNamespaceKeyFunc, store)
	return nil
}
```

### Why is this needed?

If we use a client-go informer to query Kubernetes resources, the memory utilization is very high for larger clusters.
This would make informers unusable for watching Kubernetes resources.

**Comment by aojea** [comment]
@MikeSpreitzer I kind of remember you talking about this topic
/sig api-machinery

**Comment by fruitbox12** [comment]
It sounds like you are looking for a way to reduce the memory usage of the shared index informer in the kubernetes/client-go library by storing the fetched resources on disk instead of in memory.

One way to achieve this would be to implement a custom store that is backed by a persistent storage solution like a database or file system, and then pass this store to the informer using the SetStore method you have suggested.

For example, you could create a custom store that implements the ThreadSafeStore interface and uses a database or file system to store and retrieve the resources. Then, you could pass this custom store to the informer using the SetStore method, which would override the default in-memory store.

It's worth noting that using a persistent store like this may introduce additional complexity and latency to your application, as reading and writing to the store will likely be slower than accessing data in memory. Additionally, you may need to implement additional logic to handle cases where the store becomes unavailable or goes offline.

**Comment by jeevanragula** [comment]
@fruitbox12 Yes, There will be additional complexities to handle with disk/other stores. But it is worth doing in a few cases.
In one of our use cases, we have to connect to multiple Kubernetes clusters to fetch the data, and storing them in memory is very costly.

Can I raise a PR to add a new method to set external Store?

**Comment by fruitbox12** [comment]
You can try using hypercore protocol to proxy data streams to be stored on a set of p2p nodes outside of the cluster aswell

**Comment by jeevanragula** [comment]
@fruitbox12 Can you please share more details?

**Comment by fruitbox12** [comment]
> @fruitbox12 Can you please share more details?

https://github.com/lanmower/hyper-ipc
this showcases the ability to stream child processes to remote nodes through websocket connections via NAT Holepunching from a Distributed Hash Table where the public key corresponds with p2p discovery. You can rework this to sit as a pod on a replica/daemonset such that each node on whatever cluster can all discovery eachother and share streams regardless of firewall rules, you can then save those streams to write to disk or volumes

**Comment by MikeSpreitzer** [comment]
Yes, I have definitely raised this issue too.  I think it is interesting.

An external store has two features that an ephemeral in-memory store does not have:
1. It might not be ephemeral --- it might already have contents at startup.
2. It has additional failure modes --- it may become partially or completely unavailable, and it may recover, all while the process is running.

**Comment by MikeSpreitzer** [comment]
As @aojea has noted, there are additional use cases for a custom Store for informers, besides using a Store based on external storage.

**Comment by aojea** [comment]
yep, we need to make a more comprehensive list, some of the feedback I heard about this turned out that were better solved in different ways, it is also not clear to me the orders of magnitude we are talking about here, Gigas, megas, ...
